### ISSUE TITLE
Hot certificate reload aborted on initial file access failure

### ISSUE THREAD

**Issue by peterpramb** [elastic/elasticsearch#63870]
**Elasticsearch version**:
```
Version: 7.9.2, Build: default/tar/d34da0ea4a966c4e49417f2da2f244e3e97b4e6e/2020-09-23T00:45:33.626720Z, JVM: 11.0.8
```

**Plugins installed**:
```
repository-s3
```

**JVM version**:
```
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

**OS version**:
```
Linux 3.10.0-1127.el7.x86_64 #1 SMP Tue Feb 18 16:39:12 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:
When the file monitor detects a change in the certificate files, it triggers a hot reload. Unfortunately due to the way how our certificate files are deployed, it will initially not be readable by the process, which is shown by this exception:
```
org.elasticsearch.ElasticsearchException: failed to initialize SSL KeyManager - not permitted to read certificate file [/path/to/server.crt]
        at org.elasticsearch.xpack.core.ssl.KeyConfig.unreadableKeyConfigFile(KeyConfig.java:98) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.PEMKeyConfig.getCertificateChain(PEMKeyConfig.java:84) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.PEMKeyConfig.createKeyManager(PEMKeyConfig.java:69) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.SSLService$SSLContextHolder.reloadSslContext(SSLService.java:703) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.SSLService$SSLContextHolder.reload(SSLService.java:697) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.SSLService.reloadSSLContext(SSLService.java:347) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.SSLConfigurationReloader.lambda$reloadConsumer$0(SSLConfigurationReloader.java:69) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.SSLConfigurationReloader$ChangeListener.onFileChanged(SSLConfigurationReloader.java:148) ~[?:?]
        at org.elasticsearch.watcher.FileWatcher$FileObserver.onFileChanged(FileWatcher.java:280) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.watcher.FileWatcher$FileObserver.checkAndNotify(FileWatcher.java:132) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.watcher.FileWatcher$FileObserver.updateChildren(FileWatcher.java:216) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.watcher.FileWatcher$FileObserver.checkAndNotify(FileWatcher.java:118) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.watcher.FileWatcher.doCheckAndNotify(FileWatcher.java:71) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.watcher.AbstractResourceWatcher.checkAndNotify(AbstractResourceWatcher.java:44) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.watcher.ResourceWatcherService$ResourceMonitor.run(ResourceWatcherService.java:179) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.threadpool.Scheduler$ReschedulingRunnable.doRun(Scheduler.java:213) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737) [elasticsearch-7.9.2.jar:7.9.2]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-7.9.2.jar:7.9.2]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.nio.file.AccessDeniedException: /path/to/server.crt
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219) ~[?:?]
        at java.nio.file.Files.newByteChannel(Files.java:370) ~[?:?]
        at java.nio.file.Files.newByteChannel(Files.java:421) ~[?:?]
        at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:420) ~[?:?]
        at java.nio.file.Files.newInputStream(Files.java:155) ~[?:?]
        at org.elasticsearch.xpack.core.ssl.CertParsingUtils.readCertificates(CertParsingUtils.java:97) ~[?:?]
```

The exception is completely valid, as the permissions of the certificate file will be fixed shortly later, so that the process will be able again to read it.

Unfortunately it seems that after throwing the exception, it simply gives up and does not try to reload the certificate file anymore at all:
```
$ openssl x509 -startdate -enddate -noout </path/to/server.crt
notBefore=Oct 11 05:21:46 2020 GMT
notAfter=Nov 10 05:21:46 2021 GMT
```
```
$ echo "quit" | openssl s_client -connect localhost:9200 2>/dev/null | \
> sed '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/!d'  | \
> openssl x509 -startdate -enddate -noout
notBefore=Oct 12 05:29:45 2019 GMT
notAfter=Nov 10 05:29:45 2020 GMT
```

The file permissions are indeed fixed shortly after, as can be seen after a manual process restart:
```
$ echo "quit" | openssl s_client -connect localhost:9200 2>/dev/null | \
> sed '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/!d'  | \
> openssl x509 -startdate -enddate -noout
notBefore=Oct 11 05:21:46 2020 GMT
notAfter=Nov 10 05:21:46 2021 GMT
```

The expected behaviour would be that the certificate file reload is continuously retried instead of giving up on the first failure.

**Steps to reproduce**:
This requires certificates with different validity periods:
1. Implement single-node instance with 'older' certificate.
2. Make the certificate file unreadable for the process.
3. Replace it with the 'newer' certificate.
4. Restore read permissions for the process.
5. Verify active certificate using `OpenSSL` (see above).
6. Restart the process.
7. Reverify active certificate using `OpenSSL` (see above).

**Comment by ywangd** [comment]
For unix like system, file permission change is a metadata change and updates `ctime`. The certificate reloading code checks last modified time (`mtime`). Hence the reload will not be triggered by just a permission change. One way to make this work is to touch the certificate file after permission change, i.e. `touch /path/to/server.crt`, to update the last modified time. 

Java does not provide a high level abstraction for retrieving `ctime`. So I doubt we would change the current reloading behaviour. But let me bring this up in team discussion and get back to you. In the meantime, could you please see if above "touch" alternative work for you. Thanks!

**Comment by peterpramb** [comment]
Good hint, thanks! A `touch` really triggers the reload again.

So the real issue (when I got it right) is that the file monitor triggers a reload in the corresponding subsystem, but is not aware (or ignores) the fact that the reload actually failed, so it can't trigger it again.

**Comment by ywangd** [comment]
> So the real issue (when I got it right) is that the file monitor triggers a reload in the corresponding subsystem, but is not aware (or ignores) the fact that the reload actually failed, so it can't trigger it again.
>
Reload was never triggered before you `touch` the certificate file. Changes to file permission do *not* trigger reload (because it does not change the file's lastModifiedTime).

**Comment by peterpramb** [comment]
I was talking about the initial reload failure. The following sequence, as described above:

1. Deploy new certificate file -> reload failed due to file permissions
2. Fix certificate file permissions -> no further reload

It's clear now that (2) will not trigger a reload as the `mtime` of the certificate file does not change.

But this does not explain why (1) triggered the reload only once, without caring about the reload result and retrying on that failure.

**Comment by ywangd** [comment]
Sorry I see what you mean now. You are right that the reload will not be retried. When reload is triggered, it first updates its record of lastModifiedTime to that of the file. It then does the actual reload. So regardless of whether the actual reload is successful or not, the lastModifiedTime record is already updated and hence will not reload again until the file gets updated again.

**Comment by peterpramb** [comment]
Ok, thanks for clarifying.

Would it be possible to change this behaviour? I would imagine that the same issue occurs when the certificate file is not atomically modified in place (not tested though).

**Comment by ywangd** [comment]
I added the `team-discuss` label to this issue so it will be discussed when the existing queue is cleared. Thanks!

**Comment by rjernst** [comment]
We discussed this issue today, and think the workaround of touching the file is reasonable. In normal circumstances, the file is modified in place, and permissions would not change, so this is an edge case. Additionally, the complexity of monitoring this from Java is too much. As Yang mentioned, there is no (generic) Java abstraction for ctime; we would need to special case Posix file attributes, or completely change the file watching infrastructure to use nio watching. The latter is something we may want to do in the future, but it is a very low priority. Given that we do not plan to make any changes here, I'm going to close this issue.
